CONTRIBUTING.md

```diff
@@ -58,6 +58,262 @@ To contribute a change proposal, use the following workflow:
     5. After you receive feedback from a reviewer, make the requested changes, commit them to your branch, and push them to your remote fork again.
     6. Once approval is given, feel free to squash & merge!
 
+## Writing tests
+
+We use table-driven tests because they allow us to test similar logic on many different cases in a way that is easy to both implement and understand. [This article](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) does a fantastic job explaining the motivation and structure of table-driven testing.
+
+Making table-driven tests in an environment built on the Cosmos SDK has some quirks to it, but overall the structure should be quite similar to what is laid out in the article linked above. We'll lay out two examples below (one that uses our format for messages and another that applies to keeper methods), each of which will hopefully be simple enough to copy-paste into a test file and use as a starting point for your test-writing in the Osmosis Core repo.
+
+### Rules of thumb for table-driven tests:
+
+1. Each test case should test one thing
+2. Each test case should be independent from one another (i.e. ideally, reordering the tests shouldn't cause them to fail)
+3. Functions should not be set as fields for a test case (this usually means that the table-driven approach is being sidestepped and that the logic in the function should probably be factored out to cover multiple/all test cases)
+
+### Example #1: [Message-Related Test](https://github.com/osmosis-labs/osmosis/blob/f0270d04bd77cc5e1c23f7913118b3c2ba737e97/x/tokenfactory/keeper/admins_test.go#L122-L181) 
+This type of test is mainly for functions that would be triggered by incoming messages (we interact directly with the message server since all other metadata is stripped from a message by the time it hits the msg_server):
+
+```
```

Comment:
nit: might be just my personal view so please feel free to ignore it if you disagree but I feel like having a link is sufficient. 

If there is a link, I usually skip code blocks in READMEs

Reply:
Since our goal is to make it as straightforward (and fast) as possible to copy-paste test templates, I personally prefer to keep the code blocks in there. This also lets us update/change/move the tests being linked to without having to worry about keeping the templates simple and relevant for copy-pasting